[유병건]Add: 위시리스트, 리뷰, 예약 추가 / 숙소리스트 수정

BG-YU opened the pull request:
## 수정 사항 간략한 한줄 요약
(여기에 수정 사항에 대한 간략한 한줄 요약/제목 작성해 주세요.)
- 위시 기능 추가
- 리뷰 기능 추가
- 예약 기능 추가
- 숙소리스트 수정
		 
## 수정 사항들 자세한 내용
(여기에 수정 사항에 대한 자세한 내용을 작성해 주세요.)
- 위시 기능 추가
  - 조회, 저장, 삭제 추가
- 리뷰 기능 추가
  - 조회, 저장, 삭제, 변경 추가
- 예약 기능 추가
  - 저장, 변경 추가
- 숙소리스트 수정
  - 유닛 테스트에 따른 로직 오류 사항 수정
 
## 기타 질문 및 특이 사항
(궁금한 사항들, 하면서 느낀 점들 공유해주실 것이 있으면 자유롭게 작성해 주세요.)

## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
- [x] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [x] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [x] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [x] 본 수정 사항들을 팀원들과 사전에 상의하였고 팀원들 모두 해당 PR에 대하여 알고 있습니다.
- [x] Git rebase와 squash를 했고 커밋 수가 하나 인것을 확인 했습니다.

lordmyshepherd commented on rooms/views.py:
```diff
@@ -148,4 +211,342 @@ def get(self, request):
+
+class Review(View):
+    ### 리뷰 보기 ###
+    @PublicUserDeco
```
프로젝트가 마무리되고 나면 https://github.com/wecode-bootcamp-korea/21-2nd-GroundBnB-backend/pull/12#discussion_r662081062 리뷰 참고해서 수정해보시면 좋을 것 같습니다~~

lordmyshepherd commented on rooms/views.py:
```diff
@@ -148,4 +211,342 @@ def get(self, request):
+class Review(View):
+    ### 리뷰 보기 ###
+    @PublicUserDeco
+    def get(self, request):
```
예를 들어서 room_id가 1번인 room의 리뷰 데이터 목록을 호출하는 엔드포인트라면,
RESTful하게 구현하려면, URI가 `/room/1/reviews?search=...` 형태가 되야 합니다.

resource(room data) 하나를 가져올 경우에는 `path parameter`를 사용해주세요!!
```suggestion
    def get(self, request, room_id):
        ...
```

lordmyshepherd commented on rooms/views.py:
```diff
@@ -148,4 +211,342 @@ def get(self, request):
+            limit      = int(request.GET.get('limit', 10))
+            user_infos = None
+            user_info  = {}
+
+            if not room_id:
+                return JsonResponse({'result':'INVALID REQUEST'}, status=400)
```
path parameter를 사용하면 line 232~233 코드도 없어도 될 것 같네요~
```suggestion
```

Commit:
Modify: main merge에 따른 conflict 해결
Modify: users app merge에 따른 로그인 데코레이터 수정
Modify: 컨벤션 수정
Modify: 상세페이지 merge에따른 conflict 수정

lordmyshepherd commented on rooms/views.py:
```diff
@@ -7,14 +10,61 @@
 class RoomFilterParam(View):
     # 프론트에서 벡엔드로 날려야 되는 파라미터
     def get(self, request):
+        cache.get_or_set('room_count', Room.objects.count(), 60 * 60)
```
👍

lordmyshepherd commented on rooms/views.py:
```diff
@@ -223,3 +286,350 @@ def get(self, request, room_id):
+
+            content  = datas['content']
+            room_id  = int(datas['room_id'])
+            group_id = int(datas['group_id']) if datas.get('group_id') else 0
```
`.get()` method를 사용하면, key값이 존재하지 않는 경우 0을 반환할 수 있습니다~

```suggestion
            group_id = int(data.get('group_id', 0))
```

lordmyshepherd commented on rooms/views.py:
```diff
@@ -223,3 +286,350 @@ def get(self, request, room_id):
+            if not room_id:
+                return JsonResponse({'MESSAGE':'INVALID REQUEST'}, status=400)
+
+            obj, created = Wish.objects.get_or_create(
+                user_id = user_id,
+                room_id = data['room_id']
+            )
+
+            if not created:
+                return JsonResponse({'MESSAGE' : 'EXISTS WISH ROOM'}, status=400)
```
단순히 중간테이블에 데이터가 있으면 생성하지 않고, 데이터가 없는 경우에는 새로 추가하는 로직에서는 아래와 같은 코드가 더 읽기 쉬운 것 같은데 어떠신가요?
`get_or_create`를 사용하나 아래와 같이 filter와 create을 각각 호출하나 쿼리 수는 차이가 없을 것 같습니다.

```suggestion
            if Wish.objects.filter( user_id = user_id, room_id = data['room_id']).exists():
                return JsonResponse({'MESSAGE' : 'EXISTS WISH ROOM'}, status=400)
            
            Wish.objects.create( user_id = user_id, room_id = data['room_id'])
            return return JsonResponse({'MESSAGE' : 'SUCCESS'}, status=201)
```

lordmyshepherd commented on rooms/views.py:
```diff
@@ -223,3 +286,350 @@ def get(self, request, room_id):
+                                            .order_by('-group_id', 'depth', 'id')[((offset - 1) * limit) : offset * limit]
+                                            
+            # 로그인한 유저가 상세페이지 댓글 모달을 눌렀을때의 경우
+            if user_id:
```
로직을 분리해서 (로그인한 유저가 상세페이지 댓글 모달을 눌렀을때의 경우와 비회원이 눌렀을 경우를 엔드포인트를 나눠야 코드가 간결해질 것 같습니다~

lordmyshepherd commented on rooms/views.py:
```diff
@@ -223,3 +286,350 @@ def get(self, request, room_id):
+            {
+                'room_id'       : wish.room.id,
+                'price'         : wish.room.price,
+                'point_average' : FOMATTER(ReviewEvaluation.objects.filter(
```
lambda 표현식을 사용해서 구현하신 FOMATTER 관련해서는 https://github.com/wecode-bootcamp-korea/21-2nd-GroundBnB-backend/pull/10#discussion_r661923565 이전에 리뷰 남겨드린 내용을 참고하시면 좋을 것 같습니다!

lordmyshepherd commented on rooms/views.py:
```diff
@@ -223,3 +286,350 @@ def get(self, request, room_id):
+            baby           = datas['baby']
+            room_id        = datas['room_id']
+
+            if not(check_in_date and check_out_date and 
+                    adult and kids and baby and room_id):
+                return JsonResponse({'MESSAGE' : 'INVALID REQUEST'}, status=400)
```
아래와 같이 더욱 하드하게 input을 검증할 수 있는 방법이 있습니다~

```suggestion
            if not isinstance(check_in_date, date):
                ..
            if not isinstance(check_out_date, date):
                ..
            if not isinstance(room_id, int):
                ...

```

lordmyshepherd commented on rooms/views.py:
```diff
@@ -223,3 +286,350 @@ def get(self, request, room_id):
+                    adult and kids and baby and room_id):
+                return JsonResponse({'MESSAGE' : 'INVALID REQUEST'}, status=400)
+
+            room_reservation_info = RoomReservation.objects.filter(
+                check_in_date  = datas['check_in_date'],
+                check_out_date = datas['check_out_date'],
+                room_id        = datas['room_id']
+            )
```
```suggestion
            room_reservations = RoomReservation.objects.filter(
                check_in_date  = datas['check_in_date'],
                check_out_date = datas['check_out_date'],
                room_id        = datas['room_id']
            )
```

lordmyshepherd reviewed (commented):
리뷰 반영은 프로젝트 끝나고 리펙토링할 때, 해주세요~
